openedx/core/djangoapps/content_libraries/models.py

```diff
@@ -127,3 +202,248 @@ def save(self, *args, **kwargs):  # lint-amnesty, pylint: disable=arguments-diff
+                        # LTI users can only auth throught LTI launches.
+                        self.user.set_unusable_password()
+                    self.user.save()
+                super().save(*args, **kwds)
+
+    def __str__(self):
+        return self.subject_id
+
+
+class LtiGradedResourceManager(models.Manager):
+    """
+    A custom manager for the graded resources model.
+    """
+
+    def upsert_from_ags_launch(self, user, block, resource_endpoint, resource_link):
+        """
+        Update or create a graded resource at AGS launch.
+        """
+        resource_id = resource_link['id']
+        resource_title = resource_link.get('title') or None
+        lineitem = resource_endpoint['lineitem']
+        lti_profile = user.contentlibraries_lti_profile
+        try:
+            resource = self.get(
+                profile=lti_profile,
+                usage_key=block.scope_ids.usage_id
+            )
+        except self.model.DoesNotExist:
+            resource = self.model(
+                profile=lti_profile,
+                usage_key=block.scope_ids.usage_id,
+            )
+        resource.resource_title = resource_title
+        resource.resource_id = resource_id
+        resource.ags_lineitem = lineitem
+        resource.save()
+        return resource
+
+    def get_from_user_id(self, user_id, **kwds):
+        """
+        Retrieve a resource for a given user id holding an lti profile.
+        """
+        try:
+            user = get_user_model().objects.get(pk=user_id)
+        except get_user_model().DoesNotExist as exc:
+            raise self.model.DoesNotExist('User specified was not found.') from exc
+        profile = getattr(user, 'contentlibraries_lti_profile', None)
+        if not profile:
+            raise self.model.DoesNotExist('User does not have a LTI profile.')
+        kwds['profile'] = profile
+        return self.get(**kwds)
+
+
+class LtiGradedResource(models.Model):
+    """
+    A content libraries resource launched through LTI with AGS enabled.
+
+    Essentially, an instance of this model represents a successful LTI AGS
+    launch.  This model links the profile that launched the resource with the
+    resource itself, allowing identifcation of the link through its usage key
+    string and user id.
+    """
+
+    objects = LtiGradedResourceManager()
+
+    profile = models.ForeignKey(
+        LtiProfile,
+        on_delete=models.CASCADE,
+        related_name='lti_resources',
+        help_text=_('The authorized LTI profile that launched the resource.'))
+
+    usage_key = models.CharField(
+        max_length=255,
+        help_text=_('The usage key string of the blockstore resource serving the '
+                    'content of this launch.'),
+    )
+
+    resource_id = models.CharField(
+        max_length=255,
+        help_text=_('The platform unique identifier of this resource in the '
+                    'platform, also known as "resource link id".'),
```

Comment:
The term "platform" here (and in a few other places) is ambiguous to me, because Open edX is frequently refered to as "the platform" both in the codebase and the community. Can we say "the **external** platform" in docstrings like this?